hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/PoolingMetricDescriptorSupplier.java

```diff
@@ -47,6 +48,16 @@
         poolPtr = pool.length - 1;
     }
 
+    PoolingMetricDescriptorSupplier(MetricDescriptorReusableData reusableData) {
+        allCreated = new ArrayList<>(reusableData.getAllCreatedEndSize() + LAST_CAPACITY_GROW);
```

Comment:
Thinking why we don't reuse `allCreated` as well and recreate it in every cycle instead. My thoughts:
- With this approach and generational GCs it is very likely that `allCreated` will be collected eventually by the minor GC. If we reuse it, it resides in the old generation and frequent resizes leads to periodic major GCs. 
- The next instances of `allCreated` after a growth will have their internal capacity closer to the real needs, while if we reused the same instance, we would have an instance with a bigger internal array. This is probably negligible though, even if with `ArrayList`'s grow policy after some time growing would be very unlikely.

Having the above said, I think it is good as is. I just wonder if there was any other reasoning behind the recreation that I don't see.

Reply:
Since we clear that is cleared every cycle, reusing may be even faster as I'm thinking now.

Reply:
I wouldn't be scared either. Under normal conditions, in every metrics collection cycle the same services will provide the same metrics and the same amount of them. 

What I'm not sure about is that if reusing `allCreated` is a good idea in the current form. `allCreated` can grow bigger than `pool` because there could be many copies of the descriptors created while the end descriptor(s) to collect are constructed and collected. This happens when we create template descriptors where we set a prefix a data structure name etc, that we reuse to create many concrete descriptors in the end with the same "inherited" attributes. If the source descriptor that we copied is not collected, it won't be recycled into the `pool`, since we recycle only what's collected, after it's collected. 

We could probably improve this if we treat `allCreated` as a container of "in-flight" descriptors that were instantiated or taken from the pool, but not yet recycled. What's recycled should disappear from this container. Then in the collector we could "mass-recycle" all descriptors from this container when switching from one metric source to the other [here](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsCollectionCycle.java#L67-L94). Actually, I believe this could significantly reduce the allocation of the descriptors and could even help in both keeping the containers smaller and eliminate reducing them. If you see allocations coming from the metrics appearing as allocation hotspots, I'd give this a try.

Reply:
I will rerun benchmark with current state of this PR an I will come back to you with the results. The allocation of growing array was visible so far in profiler results, mostly because the large arrays are allocated outside TLAB, and such an allocation is instantly gathered by async-profiler in alloc mode.

If I find something more to improve I will create separate PR for that.